README.rst
dhis2.py
########

|Latest version| |Downloads| |Build| |BuildWin| |Coverage| |LGTM| |CodeClimate|

A Python library for `DHIS2 <https://dhis2.org>`_ wrapping `requests <http://docs.python-requests.org/en/master/user/quickstart/>`_
for **general-purpose API interaction** with DHIS2. It attempts to be **useful for any data/metadata import and export tasks**
including various utilities like file loading, UID generation and logging. A strong focus is on JSON.

Supported and tested on Linux/macOS, Windows and DHIS2 versions >= 2.25. Python 3.6+ is required.

.. contents::
.. section-numbering::


Installation
=============

Python 3.6+ is required.

.. code:: bash

    pip install dhis2.py

For instructions on installing Python / pip for your operating system see `realpython.com/installing-python <https://realpython.com/installing-python>`_.


Quickstart
==========

Create an ``Api`` object:

.. code:: python

    from dhis2 import Api

    api = Api('play.dhis2.org/demo', 'admin', 'district')

Then run requests on it:

.. code:: python

    r = api.get('organisationUnits/Rp268JB6Ne4', params={'fields': 'id,name'})

    print(r.json())
    # { "name": "Adonkia CHP", "id": "Rp268JB6Ne4" }

    r = api.post('metadata', json={'dataElements': [ ... ] })
    print(r.status_code) # 200


- ``api.get()``
- ``api.post()``
- ``api.put()``
- ``api.patch()``
- ``api.delete()``

see below for more methods.

They all return a *Response* object from `requests <http://docs.python-requests.org/en/master/user/quickstart/>`_
except noted otherwise. This means methods and attributes are equally available
(e.g. ``Response.url``, ``Response.text``, ``Response.status_code`` etc.).

Usage
=====

Api instance creation
-----------------------

Authentication in code
^^^^^^^^^^^^^^^^^^^^^^

Create an API object

.. code:: python

    from dhis2 import Api

    api = Api('play.dhis2.org/demo', 'admin', 'district')

optional arguments:

- ``api_version``: DHIS2 API version
- ``user_agent``: submit your own User-Agent header. This is useful if you need to parse e.g. Nginx logs later.


Authentication from file
^^^^^^^^^^^^^^^^^^^^^^^^^

Load from a auth JSON file in order to not store credentials in scripts.
Must have the following structure:

.. code:: json

    {
      "dhis": {
        "baseurl": "http://localhost:8080",
        "username": "admin",
        "password": "district"
      }
    }

.. code:: python

    from dhis2 import Api

    api = Api.from_auth_file('path/to/auth.json', api_version=29, user_agent='myApp/1.0')


If no file path is specified, it tries to find a file called ``dish.json`` in:

1. the ``DHIS_HOME`` environment variable
2. your Home folder


Get info about the DHIS2 instance
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

API version as a string:

.. code:: python

    print(api.version)
    # '2.30'

API version as an integer:

.. code:: python

    print(api.version_int)
    # 30

API revision / build:

.. code:: python

    print(api.revision)
    # '17f7f0b'

API URL:

.. code:: python

    print(api.api_url)
    # 'https://play.dhis2.org/demo/api/30'

Base URL:

.. code:: python

    print(api.base_url)
    # 'https://play.dhis2.org/demo'

system info (this is persisted across the session):

.. code:: python

    print(api.info)
    # {
    #   "lastAnalyticsTableRuntime": "11 m, 51 s",
    #   "systemId": "eed3d451-4ff5-4193-b951-ffcc68954299",
    #   "contextPath": "https://play.dhis2.org/2.30",
    #   ...


Getting things
--------------

Normal method: ``api.get()``, e.g.

.. code:: python

    r = api.get('organisationUnits/Rp268JB6Ne4', params={'fields': 'id,name'})
    data = r.json()

Parameters:

- `timeout`: to override the timeout value (default: 5 seconds) in order to prevent the client to wait indefinitely on a server response.


Paging
^^^^^^

Paging for larger GET requests via ``api.get_paged()``

Two possible ways:

a) Process every page as they come in:

.. code:: python

    for page in api.get_paged('organisationUnits', page_size=100):
        print(page)
        # { "organisationUnits": [ {...}, {...} ] } (100 organisationUnits)

b) Load all pages before proceeding (this may take a long time) - to do this, do not use ``for`` and add ``merge=True``:

.. code:: python

    all_pages = api.get_paged('organisationUnits', page_size=100, merge=True):
    print(all_pages)
    # { "organisationUnits": [ {...}, {...} ] } (all organisationUnits)

*Note:* Returns directly a JSON object, not a requests.Response object unlike normal GETs.


SQL Views
^^^^^^^^^^

Get SQL View data as if you'd open a CSV file, optimized for larger payloads, via ``api.get_sqlview()``

.. code:: python

    # poll a sqlView of type VIEW or MATERIALIZED_VIEW:
    for row in api.get_sqlview('YOaOY605rzh', execute=True, criteria={'name': '0-11m'}):
        print(row)
        # {'code': 'COC_358963', 'name': '0-11m'}

    # similarly, poll a sqlView of type QUERY:
    for row in api.get_sqlview('qMYMT0iUGkG', var={'valueType': 'INTEGER'}):
        print(row)

    # if you want a list directly, cast it to a ``list`` or add ``merge=True``:
    data = list(api.get_sqlview('qMYMT0iUGkG', var={'valueType': 'INTEGER'}))
    # OR
    # data = api.get_sqlview('qMYMT0iUGkG', var={'valueType': 'INTEGER'}, merge=True)

*Note:* Returns directly a JSON object, not a requests.response object unlike normal GETs.

Beginning of 2.26 you can also use normal filtering on sqlViews. In that case, it's recommended
to use the ``stream=True`` parameter of the ``Dhis.get()`` method.



GET other content types
^^^^^^^^^^^^^^^^^^^^^^^

Usually defaults to JSON but you can get other file types:

.. code:: python

    r = api.get('organisationUnits/Rp268JB6Ne4', file_type='xml')
    print(r.text)
    # <?xml version='1.0' encoding='UTF-8'?><organisationUnit ...

    r = api.get('organisationUnits/Rp268JB6Ne4', file_type='pdf')
    with open('/path/to/file.pdf', 'wb') as f:
        f.write(r.content)



Updating / deleting things
--------------------------

Normal methods:

* ``api.post()``
* ``api.put()``
* ``api.patch()``
* ``api.delete()``


Post partitioned payloads
^^^^^^^^^^^^^^^^^^^^^^^^^^

If you have such a large payload (e.g. metadata imports) that you frequently get a HTTP Error:
``413 Request Entity Too Large`` response e.g. from Nginx you might benefit from using
the following method that splits your payload in partitions / chunks and posts them one-by-one.
You define the amount of elements in each POST by specifying a number in ``thresh`` (default: ``1000``).

Note that it is only possible to submit one key per payload (e.g. ``dataElements`` only, not additionally ``organisationUnits`` in the same payload).

``api.post_partitioned()``

.. code:: python
    
    import json
    
    data = {
        "organisationUnits": [
            {...},
            {...} # very large number of org units
        ]
    {
    for response in api.post_partitioned('metadata', json=data, thresh=5000):
        text = json.loads(response.text)
        print('[{}] - {}'.format(text['status'], json.dumps(text['stats'])))


Multiple params with same key
-----------------------------

If you need to pass multiple parameters to your request with the same key, you may submit as a list of tuples instead when e.g.:

.. code:: python

    r = api.get('dataValueSets', params=[
            ('dataSet', 'pBOMPrpg1QX'), ('dataSet', 'BfMAe6Itzgt'),
            ('orgUnit', 'YuQRtpLP10I'), ('orgUnit', 'vWbkYPRmKyS'),
            ('startDate', '2013-01-01'), ('endDate', '2013-01-31')
        ]
    )

alternatively:

.. code:: python

    r = api.get('dataValueSets', params={
        'dataSet': ['pBOMPrpg1QX', 'BfMAe6Itzgt'],
        'orgUnit': ['YuQRtpLP10I', 'vWbkYPRmKyS'],
        'startDate': '2013-01-01',
        'endDate': '2013-01-31'
    })


Utilities
---------

Load JSON file
^^^^^^^^^^^^^^^^^

.. code:: python

    from dhis2 import load_json

    json_data = load_json('/path/to/file.json')
    print(json_data)
    # { "id": ... }


Load CSV file
^^^^^^^^^^^^^^^^

Via a Python generator:

.. code:: python

    from dhis2 import load_csv

    for row in load_csv('/path/to/file.csv'):
        print(row)
        # { "id": ... }

Via a normal list, loaded fully into memory:

.. code:: python

    data = list(load_csv('/path/to/file.csv'))

Generate UID
^^^^^^^^^^^^

Create a DHIS2 UID:

.. code:: python

    uid = generate_uid()
    print(uid)
    # 'Rp268JB6Ne4'

To create a list of 1000 UIDs:

.. code:: python

    uids = [generate_uid() for _ in range(1000)]


Validate UID
^^^^^^^^^^^^

Check if something is a valid DHIS2 UID:

.. code:: python

    uid = 'MmwcGkxy876'
    print(is_valid_uid(uid))
    # True

    uid = 25329
    print(is_valid_uid(uid))
    # False

    uid = 'MmwcGkxy876 '
    print(is_valid_uid(uid))
    # False


Clean an object
^^^^^^^^^^^^^^^^

Useful for deep-removing certain keys in an object,
e.g. remove all sharing by recursively removing all ``user`` and ``userGroupAccesses`` fields.

.. code:: python

    from dhis2 import clean_obj

    metadata = {
        "dataElements": [
            {
                "name": "ANC 1st visit",
                "id": "fbfJHSPpUQD",
                "publicAccess": "rw------",
                "userGroupAccesses": [
                    {
                        "access": "r-r-----",
                        "userGroupUid": "Rg8wusV7QYi",
                        "displayName": "HIV Program Coordinators",
                        "id": "Rg8wusV7QYi"
                    },
                    {
                        "access": "rwr-----",
                        "userGroupUid": "qMjBflJMOfB",
                        "displayName": "Family Planning Program",
                        "id": "qMjBflJMOfB"
                    }
                ]
            }
        ],
        "dataSets": [
            {
                "name": "ART monthly summary",
                "id": "lyLU2wR22tC",
                "publicAccess": "rwr-----",
                "userGroupAccesses": [
                    {
                        "access": "r-rw----",
                        "userGroupUid": "GogLpGmkL0g",
                        "displayName": "_DATASET_Child Health Program Manager",
                        "id": "GogLpGmkL0g"
                    }
                ]
            }
        ]
    }


    cleaned = clean_obj(metadata, ['userGroupAccesses', 'publicAccess'])
    pretty_json(cleaned)

Which would eventually recursively remove all keys matching to ``userGroupAccesses`` or ``publicAccess``:

.. code:: json

    {
      "dataElements": [
        {
          "name": "ANC 1st visit",
          "id": "fbfJHSPpUQD"
        }
      ],
      "dataSets": [
        {
          "name": "ART monthly summary",
          "id": "lyLU2wR22tC"
        }
      ]
    }


Print pretty JSON
^^^^^^^^^^^^^^^^^

Print easy-readable JSON objects with colors, utilizes `Pygments <http://pygments.org/>`_.

.. code:: python

    from dhis2 import pretty_json

    obj = {"dataElements": [{"name": "Accute Flaccid Paralysis (Deaths < 5 yrs)", "id": "FTRrcoaog83", "aggregationType": "SUM"}]}
    pretty_json(obj)

... prints (in a terminal it will have colors):

.. code:: json

    {
      "dataElements": [
        {
          "aggregationType": "SUM",
          "id": "FTRrcoaog83",
          "name": "Accute Flaccid Paralysis (Deaths < 5 yrs)"
        }
      ]
    }


Check import response
^^^^^^^^^^^^^^^^^^^^^^

Check the importSummary response from e.g. `/api/dataValues` or `/api/metadata` import. Returns true if import went well,
false if there are ignored values or the status reports not a OK or SUCCESS.
This can be useful if the response returns a 200 OK but the summary reports ignored data.

.. code:: python

    from dhis2 import import_response_ok

    # response as e.g. from response = api.post('metadata', data=payload).json()
    response = {
        "description": "The import process failed: java.lang.String cannot be cast to java.lang.Boolean",
        "importCount": {
            "deleted": 0,
            "ignored": 1,
            "imported": 0,
            "updated": 0
        },
        "responseType": "ImportSummary",
        "status": "WARNING"
    }

    import_successful = import_response_ok(response)
    print(import_successful)
    # False

Logging
-------

Logging utilizes `logzero <https://github.com/metachris/logzero>`_.

- Color output depending on log level
- DHIS2 log format including the line of the caller
- optional ``logfile=`` specifies a rotating log file path (20 x 10MB files)


.. code:: python

    from dhis2 import setup_logger, logger

    setup_logger(logfile='/var/log/app.log')

    logger.info('my log message')
    logger.warning('missing something')
    logger.error('something went wrong')
    logger.exception('with stacktrace')

::

    * INFO  2018-06-01 18:19:40,001  my log message [script:86]
    * ERROR  2018-06-01 18:19:40,007  something went wrong [script:87]

Use ``setup_logger(include_caller=False)`` if you want to remove ``[script:86]`` from logs.

Exceptions
----------

There are two exceptions:

- ``RequestException``: DHIS2 didn't like what you requested. See the exception's ``code``, ``url`` and ``description``.
- ``ClientException``: Something didn't work with the client not involving DHIS2.

They both inherit from ``Dhis2PyException``.


Examples
========

* Real-world script examples can be found in the ``examples`` folder.
* dhis2.py is used in `dhis2-pk <https://github.com/davidhuser/dhis2-pk>`_ (dhis2-pocket-knife)

Changelog
==========

Versions `changelog <https://github.com/davidhuser/dhis2.py/blob/master/CHANGELOG.rst>`_

Contribute
==========

Feedback welcome!

- Add `issue <https://github.com/davidhuser/dhis2.py/issues/new>`_
- Install the dev environment (see below)
- Fork, add changes to *master* branch, ensure tests pass with full coverage and add a Pull Request

.. code:: bash

    pip install pipenv
    git clone https://github.com/davidhuser/dhis2.py
    cd dhis2.py
    pipenv install --dev
    pipenv run tests

    # install pre-commit hooks
    pipenv run pre-commit install

    # run type annotation check
    pipenv run mypy dhis2

    # run flake8 style guide enforcement
    pipenv run flake8

License
=======

dhis2.py's source is provided under MIT license.
See LICENCE for details.

* Copyright (c), 2020, David Huser


.. |Latest version| image:: https://img.shields.io/pypi/v/dhis2.py.svg?label=PyPi
   :target: https://pypi.org/project/dhis2.py
   :alt: PyPi version
   
.. |Downloads| image:: https://static.pepy.tech/badge/dhis2-py/month
   :target: https://pepy.tech/project/dhis2.py
   :alt: Downloads

.. |Build| image:: https://img.shields.io/circleci/project/github/davidhuser/dhis2.py/master.svg?label=Linux%20build
   :target: https://circleci.com/gh/davidhuser/dhis2.py
   :alt: CircleCI build


CHANGELOG.rst
=========
CHANGELOG
=========

2.3.0
-----
- Feat: ``import_response_ok()`` to check on the importSummary of a DHIS2 import request for data values, events, metadata.

2.2.1
-----
- Fix: no default ``timeout`` parameter value (previous: 5 seconds) in ``Api.get()``. Implementations are advised to set this per connection (see requests documentation).


2.2.0
-----
- Feat: allow a ``timeout`` parameter in ``Api.get()`` to prevent requests from waiting indefinitely on a response (see `here <https://docs.python-requests.org/en/master/user/quickstart/#timeouts>`_)


2.1.2
-----
- Fix: set ``python_requires`` correctly in setup.py

2.1.1
-----
- Chore: drop `black` from dev-packages

2.1.0
-----
- Chore: Drop Python 2.7 and 3.5 support, Python 3.6+ is required
- Chore: use type annotations, flake8, mypy, and black

2.0.2
-----
- Chore: Maintenance release for documentation

2.0.1
-----
- Chore: Maintenance release for documentation

2.0.0
-----
- Chore: ``Dhis`` class was renamed to ``Api``
- Chore: ``APIException`` was renamed to ``RequestException``
- Feat: ``generate_uids()`` was completely removed. Use ``generate_uid()`` for a single UID, not requiring an existing ``Api`` instance.

1.8.0
-----
- Feat: allow ``delete()`` to have payload

1.7.3
------
- Chore: update *requests* to 2.20.0 due to `CVE-2018-18074 <https://nvd.nist.gov/vuln/detail/CVE-2018-18074>`_

1.7.2
------
- Chore: pin test dependencies, CI tests

1.7.1
------
- Feat: allow ``delete()`` to have params

1.7.0
------
- Feat: ``clean_obj()`` to recursively remove e.g. ``userGroupAccesses`` keys from an object
- Chore: require ``six`` as a dependency

1.6.2
-----
- Chore: various minor bug fixes and clean up

1.6.1
-----
- Feat: argument ``local`` to ``generate_uids()`` to create UIDs locally (no network calls to DHIS2)
- Feat: ``pretty_json()`` to print easy-readable JSON

1.5.3
------
- Feat: partitioned payloads with ``.post_partitioned()`` to split large payloads into smaller ones

1.5.2
-----
- Feat: allow to use ``https://`` scheme and *no scheme* for localhost urls (e.g. ``localhost:8080``)
- Feat: allow to use ``json`` alongside ``data`` argument in Dhis ``.get`` / ``.post`` / ``.put`` / ``.patch`` to standardize with ``requests``

1.5.1
-----
- Feat: ``setup_logger()`` to choose if caller and line of script should be included - e.g. ``[script:123]``
- Chore: CI testing fixes

1.5.0
-----
- Feat: allow list of tuples as params to HTTP requests - `#4 <https://github.com/davidhuser/dhis2.py/issues/4>`_
- Fix: ``get_paged.page_size`` validation
- Feat: this changelog

1.4.x
-----
- Feat: ``merge`` support for ``get_sqlview()``
- Feat: enable paging of events


dhis2/api.py
# -*- coding: utf-8 -*-

"""
dhis2.api
~~~~~~~~~

This module implements DHIS2 API operations via the Api class.
"""

import codecs
from contextlib import closing
from itertools import chain
from typing import Union, Optional, Generator, List, Any, Iterator

from urllib.parse import urlparse, urlunparse

import requests
from csv import DictReader

from .exceptions import ClientException, RequestException
from .utils import load_json, partition_payload, search_auth_file, version_to_int


class Api(object):
    """A Python interface to the DHIS2 API

    Example usage:

    from dhis2 import Api

    api = Api('play.dhis2.org/demo', 'admin', 'district')

    """

    def __init__(
        self,
        server: str,
        username: str,
        password: str,
        api_version: Union[int, str] = None,
        user_agent: str = None,
    ) -> None:
        """

        :param server: baseurl, e.g. 'play.dhis2.org/demo'
        :param username: DHIS2 username
        :param password: DHIS2 password
        :param api_version: optional, creates a url like /api/29/schemas
        :param user_agent: optional, add user-agent to header. otherwise it uses requests' user-agent.
        """
        (
            self._base_url,
            self._api_version,
            self._info,
            self._version,
            self._version_int,
            self._revision,
        ) = (None,) * 6

        self.base_url = server
        self.api_version = api_version

        self.session = requests.Session()
        self.username = username
        self.session.auth = (self.username, password)
        if user_agent:
            self.session.headers["user-agent"] = user_agent

    def get_base_url(self) -> Optional[str]:
        return self._base_url

    def set_base_url(self, server: str) -> None:
        if "/api" in server:
            raise ClientException("Do not include /api/ in the DHIS2 `server` argument")

        server = server.strip()

        is_local = "localhost" in server or "127.0.0.1" in server
        has_scheme = "://" in server

        # add http / https schemes when missing
        if is_local and not has_scheme:
            url = "http://{}".format(server)
        elif not is_local and not has_scheme:
            url = "https://{}".format(server)
        else:
            url = server

        o = urlparse(url)
        self._base_url = "{}".format(
            urlunparse((o.scheme, o.netloc, o.path, "", "", ""))
        )

    def get_api_version(self) -> Optional[int]:
        return self._api_version

    def set_api_version(self, number: Union[str, int]) -> None:
        if number:
            try:
                i = int(number)
                if i < 25:
                    raise ValueError
            except ValueError:
                raise ClientException(
                    "`api_version` must be 25 or greater: {}".format(number)
                )
            else:
                self._api_version = i
        else:
            self._api_version = None

    def get_api_url(self) -> str:
        if self._api_version:
            return "{}/api/{}".format(self._base_url, self._api_version)
        else:
            return "{}/api".format(self._base_url)

    def get_info(self) -> dict:
        if not self._info:
            self._info = self.get("system/info").json()
        return self._info

    def get_version(self) -> str:
        return self._version if self._version else self.info["version"]

    def get_revision(self) -> str:
        return self._revision if self._revision else self.info["revision"]

    def get_version_int(self) -> Optional[int]:
        if not self._version_int:
            self._version_int = version_to_int(self.version)  # type: ignore
        return self._version_int

    # using property class to allow for type hinting of property (instead of @property)
    base_url = property(get_base_url, set_base_url)
    api_version = property(get_api_version, set_api_version)
    api_url = property(get_api_url)
    info = property(get_info)
    version = property(get_version)
    revision = property(get_revision)
    version_int = property(get_version_int)

    def __str__(self):
        s = (
            "DHIS2 Base URL: '{}'\n"
            "API URL: '{}'\n"
            "Username: '{}'".format(self.base_url, self.api_url, self.username)
        )
        return s

    @classmethod
    def from_auth_file(
        cls,
        location: str = None,
        api_version: Union[int, str] = None,
        user_agent: str = None,
    ) -> "Api":
        """
        Alternative constructor to load from JSON file.
        If auth_file_path is not specified, it tries to find `dish.json` in:
        - DHIS_HOME
        - Home folder
        :param location: authentication file path
        :param api_version: see Api
        :param user_agent: see Api
        :return: Api instance
        """
        location = search_auth_file() if not location else location

        a = load_json(location)
        try:
            section = a["dhis"]
            baseurl = section["baseurl"]
            username = section["username"]
            password = section["password"]
            assert all([baseurl, username, password])
        except (KeyError, AssertionError):
            raise ClientException("Auth file found but not valid: {}".format(location))
        else:
            return cls(
                baseurl,
                username,
                password,
                api_version=api_version,
                user_agent=user_agent,
            )

    @staticmethod
    def _validate_response(response: requests.Response) -> requests.Response:
        """
        Return response if ok, raise RequestException if not ok
        :param response: requests.response object
        :return: requests.response object
        """
        try:
            response.raise_for_status()
        except requests.RequestException:
            raise RequestException(
                code=response.status_code, url=response.url, description=response.text
            )
        else:
            return response

    @staticmethod
    def _validate_request(
        endpoint: str,
        file_type: str = "json",
        data: dict = None,
        params: Union[dict, List[tuple]] = None,
    ) -> None:
        """
        Validate request before calling API
        :param endpoint: API endpoint
        :param file_type: file type requested
        :param data: payload
        :param params: HTTP parameters
        """
        if not isinstance(endpoint, str) or endpoint.strip() == "":
            raise ClientException("Must submit `endpoint` for DHIS2 API")
        if not isinstance(file_type, str) or file_type.lower() not in (
            "json",
            "csv",
            "xml",
            "pdf",
            "xlsx",
        ):
            raise ClientException("Invalid file_type: {}".format(file_type))
        if params:
            if not isinstance(params, (dict, list)):
                raise ClientException(
                    "`params` must be a dict or list of tuples, not {}".format(
                        params.__class__.__name__  # type: ignore
                    )
                )
            if isinstance(params, list) and not all(
                [isinstance(elem, tuple) for elem in params]
            ):
                raise ClientException("`params` list must all be tuples")
        if data and not isinstance(data, dict):
            raise ClientException(
                "`data` must be a dict, not {}".format(data.__class__.__name__)  # type: ignore
            )

    def _make_request(
        self, method: str, endpoint: str, **kwargs: Any
    ) -> requests.Response:
        """
        Do the actual request with supplied HTTP method
        :param method: HTTP method
        :param endpoint: DHIS2 API endpoint
        :param kwargs: keyword args
        :return: response if ok, RequestException if not
        """
        if isinstance(kwargs.get("file_type"), str):
            file_type = kwargs["file_type"].lower()
        else:
            file_type = "json"
        params = kwargs.get("params")

        data = kwargs.get("data", kwargs.get("json", None))
        timeout = kwargs.get("timeout")

        url = "{}/{}".format(self.api_url, endpoint)
        self._validate_request(endpoint, file_type, data, params)

        if method == "get":
            stream = kwargs.get("stream", False)
            url = "{}.{}".format(url, file_type)
            r = self.session.get(url, params=params, stream=stream, timeout=timeout)

        elif method == "post":
            r = self.session.post(url=url, json=data, params=params, timeout=timeout)

        elif method == "put":
            r = self.session.put(url=url, json=data, params=params, timeout=timeout)

        elif method == "patch":
            r = self.session.patch(url=url, json=data, params=params, timeout=timeout)

        elif method == "delete":
            r = self.session.delete(url=url, params=params, timeout=timeout)

        else:
            raise ClientException("Non-supported HTTP method: {}".format(method))

        return self._validate_response(r)

    def get(
        self,
        endpoint: str,
        file_type: str = "json",
        params: Union[dict, List[tuple]] = None,
        stream: bool = False,
        timeout: int = None
    ) -> requests.Response:
        """
        GET from DHIS2
        :param endpoint: DHIS2 API endpoint
        :param file_type: DHIS2 API File Type (json, xml, csv), defaults to JSON
        :param params: HTTP parameters
        :param stream: use requests' stream parameter
        :param timeout: request timeout in seconds
        :return: requests.Response object
        """
        return self._make_request(
            "get", endpoint, params=params, file_type=file_type, stream=stream, timeout=timeout
        )

    def post(
        self,
        endpoint: str,
        json: dict = None,
        params: Union[dict, List[tuple]] = None,
        **kwargs: Any
    ) -> requests.Response:
        """POST to DHIS2
        :param endpoint: DHIS2 API endpoint
        :param json: HTTP payload
        :param params: HTTP parameters
        :return: requests.Response object
        """
        json = kwargs["data"] if "data" in kwargs else json
        return self._make_request("post", endpoint, data=json, params=params)

    def put(
        self,
        endpoint: str,
        json: dict = None,
        params: Union[dict, List[tuple]] = None,
        **kwargs: Any
    ) -> requests.Response:
        """
        PUT to DHIS2
        :param endpoint: DHIS2 API endpoint
        :param json: HTTP payload
        :param params: HTTP parameters
        :return: requests.Response object
        """
        json = kwargs["data"] if "data" in kwargs else json
        return self._make_request("put", endpoint, data=json, params=params)

    def patch(
        self,
        endpoint: str,
        json: dict = None,
        params: Union[dict, List[tuple]] = None,
        **kwargs: Any
    ) -> requests.Response:
        """
        PATCH to DHIS2
        :param endpoint: DHIS2 API endpoint
        :param json: HTTP payload
        :param params: HTTP parameters (dict)
        :return: requests.Response object
        """
        json = kwargs["data"] if "data" in kwargs else json
        return self._make_request("patch", endpoint, data=json, params=params)

    def delete(
        self,
        endpoint: str,
        json: dict = None,
        params: Union[dict, List[tuple]] = None,
        **kwargs: Any
    ) -> requests.Response:
        """
        DELETE from DHIS2
        :param endpoint: DHIS2 API endpoint
        :param json: HTTP payload
        :param params: HTTP parameters (dict)
        :return: requests.Response object
        """
        json = kwargs["data"] if "data" in kwargs else json
        return self._make_request("delete", endpoint, data=json, params=params)

    def get_paged(
        self,
        endpoint: str,
        params: Union[dict, List[tuple]] = None,
        page_size: Union[int, str] = 50,
        merge: bool = False,
    ) -> Union[Generator[dict, dict, None], dict]:
        """
        GET with paging (for large payloads).
        :param page_size: how many objects per page
        :param endpoint: DHIS2 API endpoint
        :param params: HTTP parameters (dict), defaults to None
        :param merge: If true, return a list containing all pages instead of one page. Defaults to False.
        :return: generator OR a normal DHIS2 response dict, e.g. {"organisationUnits": [...]}
        """
        try:
            if not isinstance(page_size, (str, int)) or int(page_size) < 1:
                raise ValueError
        except ValueError:
            raise ClientException("page_size must be > 1")

        params = {} if not params else params
        if "paging" in params:
            raise ClientException(
                "Can't set paging manually in `params` when using `get_paged`"
            )
        params["pageSize"] = page_size  # type: ignore
        params["page"] = 1  # type: ignore
        params["totalPages"] = True  # type: ignore

        collection = endpoint.split("/")[
            0
        ]  # only use e.g. events when submitting events/query as endpoint

        def page_generator() -> Generator[dict, dict, None]:
            """Yield pages"""
            page = self.get(endpoint=endpoint, file_type="json", params=params).json()
            page_count = page["pager"]["pageCount"]
            yield page

            while page["pager"]["page"] < page_count:
                params["page"] += 1  # type: ignore
                page = self.get(
                    endpoint=endpoint, file_type="json", params=params
                ).json()
                yield page

        if not merge:
            return page_generator()
        else:
            data = []
            for p in page_generator():
                data.append(p[collection])
            return {collection: list(chain.from_iterable(data))}

    def get_sqlview(
        self,
        uid: str,
        execute: bool = False,
        var: dict = None,
        criteria: dict = None,
        merge: bool = False,
    ) -> Union[Generator, List[dict]]:
        """
        GET SQL View data
        :param uid: sqlView UID
        :param execute: materialize sqlView before downloading its data
        :param var: for QUERY types, a dict of variables to query the sqlView
        :param criteria: for VIEW / MATERIALIZED_VIEW types, a dict of criteria to filter the sqlView
        :param merge: If true, return a list containing all pages instead of one page. Defaults to False.
        :return: a list OR generator where __next__ is a 'row' of the SQL View
        """
        params = {}
        sqlview_type = (
            self.get("sqlViews/{}".format(uid), params={"fields": "type"})
            .json()
            .get("type")
        )
        if sqlview_type == "QUERY":
            if not isinstance(var, dict):
                raise ClientException(
                    "Use a dict to submit variables: e.g. var={'key1': 'value1', 'key2': 'value2'}"
                )
            var = ["{}:{}".format(k, v) for k, v in var.items()]  # type: ignore
            params["var"] = var  # type: ignore
            if execute:
                raise ClientException(
                    "SQL view of type QUERY, no view to create (no execute=True)"
                )

        else:  # MATERIALIZED_VIEW / VIEW
            if criteria:
                if not isinstance(criteria, dict):
                    raise ClientException(
                        "Use a dict to submit criteria: { 'col1': 'value1', 'col2': 'value2' }"
                    )
                criteria = ["{}:{}".format(k, v) for k, v in criteria.items()]  # type: ignore
                params["criteria"] = criteria  # type: ignore

            if execute:  # materialize
                self.post("sqlViews/{}/execute".format(uid))

        def page_generator() -> Generator[dict, dict, None]:
            with closing(
                self.get(
                    "sqlViews/{}/data".format(uid),
                    file_type="csv",
                    params=params,
                    stream=True,
                )
            ) as r:
                # do not need to use unicodecsv.DictReader as data comes in bytes already
                reader = DictReader(
                    codecs.iterdecode(r.iter_lines(), "utf-8"),
                    delimiter=",",
                    quotechar='"',
                )
                for row in reader:
                    yield row

        if not merge:
            return page_generator()
        else:
            return list(page_generator())

    def post_partitioned(
        self,
        endpoint: str,
        json: dict,
        params: Union[dict, List[tuple]] = None,
        thresh: int = 1000,
    ) -> Iterator[requests.Response]:
        """
        Post a payload in chunks to prevent 'Request Entity Too Large' Timeout errors
        :param endpoint: the API endpoint to use
        :param json: payload dict
        :param params: request parameters
        :param thresh: the maximum amount to partition into
        :return: generator where __next__ is a requests.Response object
        """

        if not isinstance(json, dict):
            raise ClientException("Parameter `json` must be a dict")
        if not isinstance(thresh, int) or thresh < 2:
            raise ClientException("`thresh` must be integer of 2 or larger")

        try:
            key = next(iter(json))  # the (only) key in the payload
        except StopIteration:
            raise ClientException("`json` is empty")
        else:
            if len(json.keys()) != 1:
                raise ClientException(
                    'Must submit exactly one key in payload - e.g. json={"dataElements": [...]"}'
                )
            if not json.get(key):
                raise ClientException("payload for key '{}' is empty".format(key))
            else:
                for data in partition_payload(data=json, key=key, thresh=thresh):
                    yield self.post(endpoint, json=data, params=params)
